include/libgit4cpp/Error.h

```diff
@@ -25,36 +25,114 @@
 #ifndef LIBGIT4CPP_ERROR_H_
 #define LIBGIT4CPP_ERROR_H_
 
-#include <stdexcept>
+#include <system_error>
 
 namespace git {
 
+class git_category_impl : public std::error_category {
+    public:
+        virtual const char* name() const noexcept;
+        virtual std::string message(int ev) const;
+};
+
+const std::error_category& git_category();
+
 /**
- * An exception class carrying an error message.
+ * An exception class carrying an error code and message.
  *
  * Error is used as the standard exception class throughout libgit4cpp.
  * It can be used directly or inherited from.
  *
  * \code
  * try
  * {
- *     throw git::Error("An error has occurred");
+ *     throw git::Error(errorcode, "An error has occurred");
+ *     // or throw git::Error("Another error occurred");
  * }
  * catch (const git::Error& e)
  * {
  *     std::cerr << e.what() << "\n";
  * }
  * \endcode
  *
+ * If no error code is specified GIT_EUSER is used.
+ *
  * \note
- * git::Error is derived from std::runtime_error. It can therefore be caught by
- * `catch (const std::exception&)`, `catch (const std::runtime_error&)`, and
- * `catch (const git::Error&)`.
+ * git::Error is derived from std::system_error. It can therefore be caught by
+ * `catch (const std::exception&)`, `catch (const std::runtime_error&)`,
+ * `catch (const std::system_error&)`, and `catch (const git::Error&)`.
  */
-class Error : public std::runtime_error
+class Error : public std::system_error
 {
 public:
-    using std::runtime_error::runtime_error;
+    using std::system_error::system_error;
+    Error(int ev)
```

Comment:
Maybe `ev` could be `error_code` or something like that. What does `ev` stand for?

Reply:
It can not be `error_code` as it is not an error code but an int :grimacing: 

I _guess_ it means 'error value' as to distinguish it from a error_code object.

Just mimicked the naming from 'upstream':

![image](https://github.com/taskolib/libgit4cpp/assets/16012374/5cad2b13-ebed-4199-8f99-11d9ea9c28bb)

Reply:
They are funny on cppreference: "Constructs with underlying *error code* `ev`"... Which just proves that naming is hard. :-D